### ISSUE TITLE
Extraction from data table is a magnitutde slower than R base matrices

### ISSUE THREAD

**Issue by melmasri** [Rdatatable/data.table#3036]
Reading data from disk and working with data is significantly faster than `R` base, extraction from data.table is much much slower than `R` base matrices. 

# Reproducibility script.

```R
## clean session
rm(list=ls())
library(data.table)
library(microbenchmark)

dt = data.table(x = runif(1000e4))      # data.table
df = data.frame(x = runif(1000e4))      # data.frame
mt = matrix( runif(1000e4), ncol=1)     # matrix
## insertion or assignment
microbenchmark(dt[,y := 2*x],  df$y <- 2*df$x , mt[,1] <- 2*mt[,1])
## extraction
microbenchmark(dt[1,1],dt$x[1], dt[['x']][1], df[1,1], df$x[1], mt[1,1])
```
# Output
```R
## insertion or assignment
> microbenchmark(dt[,y := 2*x],  df$y <- 2*df$x , mt[,1] <- 2*mt[,1])
Unit: milliseconds
                   expr      min       lq     mean   median       uq       max
   dt[, `:=`(y, 2 * x)] 28.62838 29.35892 37.14976 29.79951 37.80667  79.92419
       df$y <- 2 * df$x 28.17611 28.78622 35.80632 29.07809 34.91636  81.94791
 mt[, 1] <- 2 * mt[, 1] 78.69023 79.87831 95.17410 87.01249 97.54382 135.62911
> ## extraction
> microbenchmark(dt[1,1],dt$x[1], dt[['x']][1], df[1,1], df$x[1], mt[1,1])
Unit: nanoseconds
         expr    min       lq      mean   median       uq    max neval
     dt[1, 1] 198587 204123.5 216862.83 205844.5 209051.5 602078   100
      dt$x[1]   6919   7628.0   9191.38   8597.5  10383.5  20940   100
 dt[["x"]][1]   5219   5992.5   7057.02   6649.0   7653.0  10220   100
     df[1, 1]  12002  13544.0  16896.79  17067.5  19033.0  39374   100
      df$x[1]   5538   6366.5   7695.67   7328.0   9137.0  11404   100
     mt[1, 1]    154    280.0    451.11    384.5    579.5   2047   100
> 
```

**Comment by franknarf1** [comment]
`dt[["x"]][1]` syntax is fine with a data.table just as with a data.frame, though if you're doing that extraction often enough that its performance matters, maybe you're approaching the problem the wrong way (eg, ought to use a matrix or do these extractions/assignments batched together)

**Comment by mattdowle** [comment]
Yes. @melmasri, your timings are so tiny (milliseconds and nanoseconds) that you're picking up the very small overhead that the `[.data.table` method has. You must scale up the data size to benchmark anything useful.

In other words `dt[1,1]` is not supposed to fast on a microbenchmark scale.  Don't use `[.data.table` for that, use one of the other methods if that's what you need.  Like as frank says,  using `[[`.

We really need #3028 implemented to explain to people why microbenchmark() causes so much trouble.

**Comment by st-pasha** [comment]
It's even worse if you consider assignment to a single cell (although data.table is on par with other data frames implementations, it loses awfully to matrix):
```
> microbenchmark(dt[1,1]<-0, dt$x[1]<-0, dt[['x']][1]<-0, df[1,1]<-0, df$x[1]<-0, mt[1,1]<-0)
Unit: nanoseconds
              expr      min         lq         mean    median        uq
     dt[1, 1] <- 0 17851497 18940511.5  27863655.44  20372754  30568036
      dt$x[1] <- 0 74572054 87539637.5 109555404.11 103324377 133275944
 dt[["x"]][1] <- 0 17619570 18675037.0  26332648.28  19561804  23516433
     df[1, 1] <- 0 26030653 28002229.0  41239656.67  31399078  51274603
      df$x[1] <- 0 17631639 18603076.0  25898066.15  20275188  24642572
     mt[1, 1] <- 0      925     4279.5      5088.22      5127      6251
```

However, given that the problem is caused by heavy amount of processing inside `[...]`, it is unlikely to be solvable.

**Comment by melmasri** [comment]
I assume my use is very specific where I repetitively extract and insert over 1000's of iterations (in an MCMC setting).   

Nonetheless, with my shallow understanding of data.table that columns are essentially lists, yet comparing to lists things are still slower in data.table. Lists and matrices are comparable. 

@mattdowle I can't push more than `runif(100e5)` on this code below with my 4GB RAM laptop. Thanks for the tips about `[[`, and yes I agree some documentation on this topic would be helpful.

```R
rm(list=ls())
library(data.table)
library(microbenchmark)

rand = runif(100e5)
dt = data.table(x = rand)      # data.table
df = data.frame(x = rand)      # data.frame
mt = matrix( rand, ncol=1)     # matrix
lt = lapply(rand, list)

> microbenchmark(dt[1,1],dt$x[1], dt[['x']][1], df[1,1], df$x[1], mt[1,1], lt[[100]][[1]], times=10000)
Unit: nanoseconds
           expr    min       lq        mean   median       uq    max neval
       dt[1, 1] 190191 219591.0 231544.9176 231460.0 237184.5 535318 10000
        dt$x[1]   7080   9216.0  11204.9970  10966.0  12615.0  37568 10000
   dt[["x"]][1]   5496   6820.5   8723.3635   8557.0   9920.5  54699 10000
       df[1, 1]  11838  15957.0  18904.3565  18477.0  21510.5  87899 10000
        df$x[1]   5779   7725.5   9673.6183   9514.5  10968.5  42741 10000
       mt[1, 1]    152    297.0    507.1780    401.0    557.0  14165 10000
 lt[[100]][[1]]    277    433.0    739.3044    637.0    885.0  11915 10000
```

**Comment by mattdowle** [comment]
@st-pasha I don't see a problem here. You look for problems and enjoy taking aim, so you see problems. Where I see different use-cases and have a generally positive outlook. You see a big number 17 million as big and "awful" relative to 925.  I see a scale of nanoseconds, translate 17 million to 0.017 seconds and combine that with i) [.data.table is not for being iterated, ii) we have low-overhead set() as well as [[.

@melmasri Correct: data.table is probably not for your case. Where matrix is suitable, use matrix.  You should only consider data.table if matrix doesn't work for you. Data.table is for the things the package description says it's for and the list of bullet points on the home page.  Although you have big vectors, the queries are selecting just one item from them. When we ask you to scale up the data we mean use it too (!) like aggregating it.

**Comment by melmasri** [comment]
Thanks for the clarification @mattdowle . I use data.table for many cases, it is just this case that I noticed a difference, now I understand it better. Thanks.

**Comment by st-pasha** [comment]
@mattdowle I see your point. 0.017s is not too bad; and there exist MUCH faster alternatives, like `set()`. But you're looking at this from the summit of your expert knowledge. Most users of data.table, however, are noobs (this is true for any product). And they would shoot themselves in the foot, given an opportunity. Should we give them this opportunity?

Here's a somewhat larger dataset (taken directly from `?set` help page):
```
> DT = as.data.table(matrix(1, nrow = 2e6, ncol = 100))
> system.time({DT[10, 10] = 100})
   user  system elapsed 
  0.457   0.338   0.796 
```
I know this is not the best way to assign a single cell. Probably the worst way. But it is also the most obvious way... 

Now, I don't know the internals of data.table well enough to understand the reason for such slowness. Maybe it is not possible to improve. But maybe it is? I do notice that the timing is roughly proportional to the size of the original dataset (so for `nrow=2e7` the time is 29.6s already for a single assignment). Is this expected? Also, if this is not possible to resolve -- is it possible to at least detect such bad use and warn the user, suggesting to use `set()` instead? After all, `data.table` likes being nice and instructive to the user.

**Comment by mattdowle** [comment]
@st-pasha  Here's that example from `?set` including its context : 
![image](https://user-images.githubusercontent.com/2779998/45240683-57a7ed00-b29e-11e8-96c3-622e7fe34481.png)

I wrote that example in `?set`, iirc. After all these years, I'm pleased I showed the result of a single `system.time()` for the total of a `for` loop running 1,000 iterations. The magnitude in difference is already shown right there. I think the comment after it is clear and should have been included (context should be included) when you quote the docs in this way.  [Aside: When you `Rprof()` using base R's `=` or `<-` on data.table, you see 90% of the time is spent in `copy()`. The `copy()` happens on each and every iteration.]

Can we do more to save beginners who don't read the documentation, don't follow links and don't search stack overflow? Sure, we can do more, and that's a good idea. But I can't be bothered when you accuse me of approaching this from the "summit of your expert knowledge".  That's derogatory and I check-out.
